providers/azure_test.go

```diff
@@ -122,98 +157,154 @@ func testAzureBackend(payload string) *httptest.Server {
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": [], "userPrincipalName": "user@windows.net" }`,
+			ExpectedEmail:           "user@windows.net",
+		},
+		{
+			Description:             "should return error when Azure backend doesn't return email information",
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": [], "userPrincipalName": null }`,
+			ExpectedError:           fmt.Errorf("unable to get email address, error: %v", errors.New("type assertion to string failed")),
+		},
+		{
+			Description:             "should return specific error when unable to get email",
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": [], "userPrincipalName": "" }`,
+			ExpectedError:           errors.New("unable to get email address"),
+		},
+		{
+			Description:             "should return error when otherMails from Azure backend is not a valid type",
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": "", "userPrincipalName": null }`,
+			ExpectedError:           fmt.Errorf("unable to get email address, error: %v", errors.New("type assertion to string failed")),
+		},
+		{
+			Description:   "should not query profile api when email is already set in session",
+			Email:         "user@windows.net",
+			ExpectedEmail: "user@windows.net",
+		},
+	}
 
-	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
-	assert.Equal(t, nil, err)
-	assert.Equal(t, "", email)
+	for _, testCase := range testCases {
+		t.Run(testCase.Description, func(t *testing.T) {
+			var (
+				b    *httptest.Server
+				host string
+			)
+			if testCase.PayloadFromAzureBackend != "" {
+				b = testAzureBackend(testCase.PayloadFromAzureBackend)
+				defer b.Close()
+
+				bURL, _ := url.Parse(b.URL)
+				host = bURL.Host
+			}
+			p := testAzureProvider(host)
+			session := CreateAuthorizedSession()
+			session.Email = testCase.Email
+			err := p.EnrichSession(context.Background(), session)
+			assert.Equal(t, testCase.ExpectedError, err)
+			assert.Equal(t, testCase.ExpectedEmail, session.Email)
+		})
+	}
 }
 
-func TestAzureProviderGetEmailAddressIncorrectOtherMails(t *testing.T) {
-	b := testAzureBackend(`{ "mail": null, "otherMails": "", "userPrincipalName": null }`)
-	defer b.Close()
-
-	bURL, _ := url.Parse(b.URL)
-	p := testAzureProvider(bURL.Host)
-
-	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
-	assert.Equal(t, "type assertion to string failed", err.Error())
-	assert.Equal(t, "", email)
-}
+func TestAzureProviderRedeem(t *testing.T) {
+	testCases := []struct {
+		Name                 string
+		RefreshToken         string
+		ExpiresOn            string
+		EmailFromIDToken     string
+		EmailFromAccessToken string
+		IsIDTokenMalformed   bool
+	}{
+		{
+			Name:             "with id_token returned",
+			EmailFromIDToken: "foo1@example.com",
+			RefreshToken:     "some_refresh_token",
+			ExpiresOn:        "2006-01-02T22:04:05Z",
```

Comment:
This feels a little weird to mock without being able to mock the session `CreatedAt` (since azure provider explicitly calls `time.Now()`) -- getting test sessions out of this where the expires date is years before the created data might get us into weird trouble in the future if functionality changes.

Can we set this to a variable set `time.Now()` + 1 hour duration and check that in assertions below?

Reply:
(I understand you are just copying this logic over -- but looking at the dated style of these Azure unit tests, this provider's test suite needs some love. So lets try to fix up oddities we catch in the diff).

Reply:
sure